CLI for the create-or-update script

phargogh opened the pull request:
This PR modifies the create-or-update script to add an `argparse`-based CLI.  For example:

```shell
# Create on prod, api key taken from CKAN_APIKEY
python create-or-update-dataset.py test_gmm.yml
```

```shell
# create on staging, api key taken from CKAN_STAGING_APIKEY
python create-or-update-dataset.py test_gmm.yml --staging
```

```shell
# create on localhost, api key taken from CKAN_LOCAL_APIKEY
python create-or-update-dataset.py test_gmm.yml --dev
```

You can also provide an API key through the CLI by passing `--apikey=...` with your string API key to the target CKAN instance.

claire-simpson commented on api-scripts/create-or-update-dataset.py:
```diff
@@ -540,5 +535,71 @@ def main(gmm_yaml_path, private=False, group=None):
-    main(sys.argv[1])
+    host, apikey, gmm_path = _ui()
+    main(host, apikey, gmm_path,
+         verify_ssl=(host.split('https://'[1].startswith('localhost'))))
```
I think there is a syntax error here, and this should be `host.split('https://')[1].startswith('localhost')` (or could it just be simply `host.startswith("https://localhost")` ?)

megannissel replied:
As Claire noted, this is definitely a syntax error. And I also think we need this check to be the opposite; as written, `verify_ssl` is being set to `True` on localhost, which causes the script to fail:
```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate (_ssl.c:1010)
```

claire-simpson commented on api-scripts/create-or-update-dataset.py:
```diff
@@ -309,17 +303,18 @@ def _get_wgs84_bbox(config):
-    session.headers.update({'Authorization': MODIFIED_APIKEY})
+    session.headers.update({'Authorization': ckan_apikey})
 
     session = requests.Session()
```
Curious why this is here twice - wouldn't this overwrite the previous `session`?

claire-simpson reviewed (changes requested):
Thanks James - very helpful! Just a couple small comments

megannissel commented on api-scripts/create-or-update-dataset.py:
```diff
@@ -37,23 +37,18 @@
-VERIFY=True
-if URL.split('https://')[1].startswith('localhost'):
-    VERIFY=False
+CKAN_HOSTS = {
```
I appreciate this mapping for the hosts and keys -- visually much cleaner!

megannissel reviewed (changes requested):
Thanks, James! Same syntax error Claire noted, plus an additional note about SSL verification for local dev.

phargogh commented:
Thanks @claire-simpson and @megannissel !  I don't know how those happened ... I want to say it was because of a merge conflict, but programmer error is more likely.  I've addressed all concerns raised so far (I think), but please let me know if you see anything else!

megannissel commented on api-scripts/create-or-update-dataset.py:
```diff
@@ -601,5 +599,6 @@ def _ui(args=None):
-    main(host, apikey, gmm_path,
-         verify_ssl=(host.split('https://'[1].startswith('localhost'))))
+
+    is_localhost = not host.split('https://')[1].startswith('localhost')
```
The `not` here and in `verify_ssl=(not is_localhost)` is putting us back where we started, I think! 
```suggestion
    is_localhost = host.split('https://')[1].startswith('localhost')
```

claire-simpson replied:
+ 1 on this - otherwise it looks good to me!

phargogh replied:
Ugh, sorry about the sloppiness here!  And thank you both for your patience

megannissel replied:
No worries at all; I have 100% been there. Somehow it always seems to be the tiny changes where silly mistakes happen most easily!

megannissel reviewed (changes requested):
Looks like we have a double negative in the localhost check!

megannissel reviewed (approved):
Looks great!
